Exact Eval

mvpatel2000 opened the pull request:
# What does this PR do?

Distributed samplers pad batches with duplicate samples in order to ensure all ranks have the same number of samples. This PR slices off duplicate samples (and adds a unit test) to ensure exact eval.

# What issue(s) does this change relate to?

[CO-2082](https://mosaicml.atlassian.net/browse/CO-2082)

[CO-2082]: https://mosaicml.atlassian.net/browse/CO-2082?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dakinggg commented on tests/trainer/test_trainer_eval.py:
```diff
@@ -52,6 +52,27 @@ def test_eval_call():
 
+@device('gpu')
+@world_size(2)
+@pytest.mark.parametrize('size', [4, 5])
```
lets just add lots of odd numbers to both of these

mvpatel2000 commented:
@dakinggg  @eracah  I've reworked parts of the inner loop. It's more complicated but required for batches which have partial data. Also added more robust tests

mvpatel2000 commented:
Thanks to Evan for teaching me about torch.split

dakinggg commented on tests/trainer/test_trainer_eval.py:
```diff
@@ -52,6 +52,28 @@ def test_eval_call():
+    expected_acc = 1 - (size // 2) / size
+    metric = trainer.state.eval_metrics['eval']['MulticlassAccuracy']
+    assert metric.compute() - expected_acc < 1e-5
+    assert metric.tp + metric.tn + metric.fp + metric.fn == size  # type: ignore
```
```suggestion
    assert metric.tp.item() + metric.fn.item()  == size  # type: ignore
```
the way torchmetrics handles `MulticlassAccuracy`, adding everything together is double counting. see https://github.com/Lightning-AI/torchmetrics/blob/17c0e9fdf636cff7299fc3d86fa5e3709b603737/src/torchmetrics/functional/classification/accuracy.py#L76-L83

mvpatel2000 commented on composer/trainer/trainer.py:
```diff
@@ -2686,6 +2696,13 @@ def _eval_loop(
+                    batch_num_samples_tensor = self.state.device.tensor_to_device(torch.tensor(rank_num_samples))
+                    dist.all_reduce(batch_num_samples_tensor, reduce_operation='SUM')
+                    batch_num_samples = batch_num_samples_tensor.item()
+                    last_batch = self.state.eval_timestamp.sample + batch_num_samples > dataset_len
```
If it's equal, there's no padding so it's fine to ignore. We only want to handle the case where there's padding.

I will update to >= because otherwise the variable is confusing, but I'm not going to add a unit test because it (correctly) no-ops in both > or >=. With >=, This will never pass:
```
padding = dist_sampler.total_size - dataset_len
if dist.get_global_rank() >= dist.get_world_size() - padding:
```

mvpatel2000 commented on composer/trainer/trainer.py:
```diff
@@ -2670,12 +2670,22 @@ def _eval_loop(
-                dataloader.sampler.set_epoch(int(self.state.timestamp.batch))
+                dist_sampler = dataloader.sampler
+                dist_sampler.set_epoch(int(self.state.timestamp.batch))
+                try:
```
Now issues warning and updated fn description -> docs should be updated

dakinggg commented on composer/trainer/trainer.py:
```diff
@@ -2571,6 +2571,13 @@ def eval(
+
+            If evaluating with multiple GPUs using a DistributedSampler with `drop_last=False`, the last
+            batch will contain duplicate samples, which may affect metrics. To avoid this, as long as
+            the dataset passed to the DistributedSampler has a length defined, duplicate samples will
```
suggest "... defined, Composer will correctly drop the duplicated samples"

dakinggg commented on composer/trainer/trainer.py:
```diff
@@ -2670,12 +2677,29 @@ def _eval_loop(
+                    except AttributeError:
+                        warnings.warn("DistributedSampler's dataset attribute does not have length. When "
+                                      '`drop_last=False`, metrics may be incorrect as DistributedSampler '
+                                      'duplicates samples to make all batches the same size. To fix this, '
```
correct me if im wrong, but it doesn't make "all batches the same size", but rather, "the dataset size divisible by the world size", right?

dakinggg commented on tests/trainer/test_trainer_eval.py:
```diff
@@ -52,6 +52,30 @@ def test_eval_call():
+    metric = trainer.state.eval_metrics['eval']['MulticlassAccuracy']
+    assert metric.compute() - expected_acc < 1e-5
+    count = metric.tp + metric.fn  # type: ignore
+    dist.all_reduce(count)
```
😢

dakinggg reviewed (commented):
Can we run one of the larger ICL eval datasets on MPT 7B and confirm that we don't see much change?

dakinggg commented on tests/trainer/test_trainer_eval.py:
```diff
@@ -52,6 +52,30 @@ def test_eval_call():
     assert trainer.state.eval_metrics['eval']['MulticlassAccuracy'].compute() != 0.0
 
 
+@world_size(1, 2)
```
could you add an FSDP parametrization to this test, just to be safe?

mvpatel2000 replied:
Will combine with ICL

mvpatel2000 replied:
Discussed offline, fine to skip

dakinggg commented on tests/trainer/test_trainer_eval.py:
```diff
@@ -52,6 +52,30 @@ def test_eval_call():
+@pytest.mark.parametrize('size', [12, 13, 14, 15, 16])
+@pytest.mark.parametrize('batch_size', [1, 2, 3, 4, 6])
+@pytest.mark.filterwarnings(r'ignore:Cannot split tensor of length.*:UserWarning')
+def test_eval_with_nondivisible_dataset(world_size, size, batch_size):
```
```suggestion
def test_eval_with_nondivisible_dataset(world_size: int, size: int, batch_size: int):
```

dakinggg commented:
Running this crashed for me when running triviaqa, at the end of the dataset.

```
Traceback (most recent call last):
  File "/workspace/llm-foundry/scripts/eval/eval.py", line 82, in <module>
    main(cfg)
  File "/workspace/llm-foundry/scripts/eval/eval.py", line 63, in main
    trainer.eval(eval_dataloader=evaluators)
  File "/usr/lib/python3/dist-packages/composer/trainer/trainer.py", line 2630, in eval
    self._eval_loop(
  File "/usr/lib/python3/dist-packages/composer/trainer/trainer.py", line 2748, in _eval_loop
    self.state.batch = data_spec.split_batch(self.state.batch,
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 91, in _default_split_batch
    return _split_mapping(batch, microbatch_size)
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 63, in _split_mapping
    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 63, in <listcomp>
    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
  File "/usr/lib/python3/dist-packages/composer/core/data_spec.py", line 63, in <dictcomp>
    return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
IndexError: list index out of range

----------End global rank 30 STDERR----------ERROR:composer.cli.launcher:Global rank 24 (PID 1079) exited with code 1
```

dakinggg commented:
Ah, QA dataset also needs a custom split batch

dakinggg commented on tests/datasets/test_in_context_learning_datasets.py:
```diff
@@ -350,6 +351,118 @@ def test_mc_task_dataloader_opt_tokenizer(dataset_uri, num_fewshot, tmp_path):
+    assert isinstance(dl.dataloader, DataLoader)  # pyright
+    batch = next(dl.dataloader._get_iterator())
+    choices_per_question = 2
+    microbatches = dl.split_batch(batch, 1)
```
nit: compute `microbatch_size` before calling `split_batch` and pass it in as an arg.
